src/audio/mixin_mixout/mixin_mixout_hifi3.c

```diff
@@ -0,0 +1,533 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Andrula Song <xiaoyuan.song@intel.com>
+
+#include <ipc4/mixin_mixout.h>
+#include <sof/common.h>
+
+#ifdef MIXIN_MIXOUT_HIFI3
+
+#if CONFIG_FORMAT_S16LE
+/* Instead of using sink->channels and source->channels, sink_channel_count and
+ * source_channel_count are supplied as parameters. This is done to reuse the function
+ * to also mix an entire stream. In this case the function is called with fake stream
+ * parameters: multichannel stream is treated as single channel and so the entire stream
+ * contents is mixed.
+ */
+static void normal_mix_channel_s16(struct audio_stream __sparse_cache *sink, uint32_t start_frame,
+				   uint32_t mixed_frames,
+				   const struct audio_stream __sparse_cache *source,
+				   uint32_t frame_count, uint16_t gain)
+{
+	uint32_t frames_to_mix, frames_to_copy, left_frames;
+	uint32_t n, nmax, i;
+	ae_int16x4 in_sample;
+	ae_int16x4 out_sample;
+	ae_int16x4 *in;
+	ae_int16x4 *out;
+	ae_valign inu = AE_ZALIGN64();
+	ae_valign outu1 = AE_ZALIGN64();
+	ae_valign outu2 = AE_ZALIGN64();
+	/* audio_stream_wrap() is required and is done below in a loop */
+	ae_int16 *dst = (ae_int16 *)sink->w_ptr + start_frame;
+	ae_int16 *src = (ae_int16 *)source->r_ptr;
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = MIN(mixed_frames - start_frame, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+	n = 0;
+
+	for (left_frames = frames_to_mix; left_frames; left_frames -= n) {
+		src = audio_stream_wrap(source, src + n);
+		dst = audio_stream_wrap(sink, dst + n);
+		/* calculate the remaining samples*/
+		nmax = audio_stream_samples_without_wrap_s16(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, nmax);
+		in = (ae_int16x4 *)src;
+		out = (ae_int16x4 *)dst;
+		inu = AE_LA64_PP(in);
+		outu1 = AE_LA64_PP(out);
+		for (i = 0; i < n; i += 4) {
```

Comment:
Please add comment where it's ensured that frames_to_mix and frame_count have multiple of four samples

Reply:
it doesn't need to be multiple of four samples, since each time I update the src and dst base on n.

Reply:
Almost forget that, I'd like to use another PR to use the frame limits API to add limits of processing limits. Thanks!